### DISCUSSION TITLE
Issue with parrallel states

### DISCUSSION THREAD

**Discussion by red-square-blue-circle** [alexzhornyak/uscxml#3] [Q&A] [answered]
I was wondering if the issue with parallel states from tklab-tud's uscxml was fixed in this fork.
I am referring to the issue [#177](https://github.com/tklab-tud/uscxml/issues/177).
here is a simple scxml file that highlights the bug: [parrallel_bug.zip](https://github.com/alexzhornyak/uscxml/files/6365550/parrallel_bug.zip)

**Comment by alexzhornyak** [comment, marked as answer]
Hi, I've tested these both parallel examples from issue #177 and **parallel_bug.scxml** with official latest uscxml and this fork. Have not discovered any errors on both branches

## Conditions:
- MSVC 2015 
- Win 7.1

## Official uscxml results
### SCXML_Parallel_NotWorking
![SCXML_Parallel_NotWorking](https://user-images.githubusercontent.com/18611095/115881583-8f1a3680-a454-11eb-8699-4d6d3cfea504.gif)


```
[Info] HTTP server listening on tcp/5080
[Info] WebSocket server listening on tcp/5081
Available Datamodels:
	ecmascript
	lua
	null
Available Invokers:
	dirmon	(DirectoryMonitor, http://uscxml.tk.informatik.tu-darmstadt.de/#dirmon)
	scxml	(uscxml, http://www.w3.org/TR/scxml, http://www.w3.org/TR/scxml/)
Available I/O Processors:
	basichttp	(http://www.w3.org/TR/scxml/#BasicHTTPEventProcessor)
	http	(http://www.w3.org/TR/scxml/#HTTPEventProcessor)
	scxml	(http://www.w3.org/TR/scxml/#SCXMLEventProcessor)
Available Elements:
	http://www.w3.org/2005/07/scxml / respond
[Info] Processing C:\Users\user\Downloads\ParallelSCXML_Examples\SCXML_Parallel_NotWorking.scxml~
[Info] Using cache from 'C:\Users\user\AppData\Local\Temp\\uscxml\4DC5AD2480DB208519070F3836BCBF54.uscxml.cache'
Microstep in config: {}
Entering: /scxml[1]
Entering: s0
Entering: p0
Entering: p1out
Entering: p1
Entering: p2out
Entering: p2
Microstep in config: {/scxml[1], s0, p0, p1out, p1, p2out, p2}
Exiting: p2
Exiting: p1
Transition: //state[@id="p1"]/transition[1]
Transition: //state[@id="p2"]/transition[1]
Entering: p11
Entering: p21
Microstep in config: {/scxml[1], s0, p0, p1out, p11, p2out, p21}
Exiting: p21
Exiting: p2out
Executable Content: //state[@id="p2out"]/onexit[1]/log[1]
[Log] "leaving state p20ut"
Exiting: p11
Exiting: p1out
Executable Content: //state[@id="p1out"]/onexit[1]/log[1]
[Log] "leaving state p1out"
Exiting: p0
Executable Content: //parallel[@id="p0"]/onexit[1]/log[1]
[Log] "leaving state parallel_state"
Exiting: s0
Transition: //state[@id="s0"]/transition[1]
Executable Content: //state[@id="s0"]/transition[1]/log[1]
[Log] "leaving s0"
Entering: pass
```

## Parallel_bug
![parallel_bug](https://user-images.githubusercontent.com/18611095/115883530-8e829f80-a456-11eb-913f-b183ea64aedf.gif)

```
[Info] HTTP server listening on tcp/5080
[Info] WebSocket server listening on tcp/5081
Available Datamodels:
	ecmascript
	lua
	null
Available Invokers:
	dirmon	(DirectoryMonitor, http://uscxml.tk.informatik.tu-darmstadt.de/#dirmon)
	scxml	(uscxml, http://www.w3.org/TR/scxml, http://www.w3.org/TR/scxml/)
Available I/O Processors:
	basichttp	(http://www.w3.org/TR/scxml/#BasicHTTPEventProcessor)
	http	(http://www.w3.org/TR/scxml/#HTTPEventProcessor)
	scxml	(http://www.w3.org/TR/scxml/#SCXMLEventProcessor)
Available Elements:
	http://www.w3.org/2005/07/scxml / respond
[Info] Processing C:\Users\user\Downloads\ParallelSCXML_Examples\parrallel_bug.scxml~
[Info] Using cache from 'C:\Users\user\AppData\Local\Temp\\uscxml\17774D1FC51C05CBA4E3887C68670427.uscxml.cache'
Microstep in config: {}
Entering: /scxml[1]
Entering: initial
Entering: parallel
Entering: parallel_atom
Microstep in config: {/scxml[1], initial, parallel, parallel_atom}
Exiting: parallel_atom
Exiting: parallel
Transition: //state[@id="parallel_atom"]/transition[1]
Entering: s0
Executable Content: //state[@id="s0"]/onentry[1]/log[1]
[Log] "s0_entry"
Executable Content: //state[@id="s0"]/onentry[1]/raise[1]
Internal Event: e
Microstep in config: {/scxml[1], initial, s0}
Exiting: s0
Exiting: initial
Transition: //state[@id="s0"]/transition[1]
Executable Content: //state[@id="s0"]/transition[1]/log[1]
[Log] "Will not be printed because it will crash before"
Entering: final
```

**Reply by alexzhornyak** [reply]
P.S. If you are able to use Windows OS, you may use [ScxmlEditor](https://github.com/alexzhornyak/ScxmlEditor-Tutorial) as testing instrument for executing scxml state charts. It has predefined UScxmlTester and QtScxmlTester. More detaily is described [here](https://github.com/alexzhornyak/ScxmlEditor-Tutorial/blob/master/Doc/DebugScxmlStateCharts.md)

**Reply by red-square-blue-circle** [reply]
thanks for the answer but after testing with the latest uscxml , i can't get it to work, I'm using uscxml-browser with parrallel_bug.uscxml and i get a Debug Assertion failed, as for you fork i can't get it to build ,first the xerces-c link seems dead , when it's fixed boost cannot be find , after including it in the cmake, some files are missing like Vld.h for example 
I'm working on Win 10 with VS 2017

**Reply by alexzhornyak** [reply]
Hi, I understood what is happening. It successfully runs on Release version but has assertion in Debug version of USCXML (official). Therefore in my fork it successefully runs on both.

## Official Debug
![image](https://user-images.githubusercontent.com/18611095/116209378-85414d80-a74a-11eb-9c42-6da4375a3dc5.png)

## This Fork Debug
![image](https://user-images.githubusercontent.com/18611095/116209604-c20d4480-a74a-11eb-8483-6a2168130aa7.png)

P.S. This fork was mostly used by me only and I made fixed solution for my needs. Give me time to update cmake to be working or I may compile libs for MSVC 2017 and put as release

**Reply by red-square-blue-circle** [reply]
I also have a fork only I use, what i was wondering is what did you did to fix this issue ? do you just use the release version and ignore the assertion ?

**Reply by alexzhornyak** [reply]
> do you just use the release version and ignore the assertion ?

I use much more previous release of the official. 
![image](https://user-images.githubusercontent.com/18611095/116217367-3c8d9280-a752-11eb-8c07-cf2f1cf1a004.png)

It was so buggy and I was tired to catch new bugs.
![image](https://user-images.githubusercontent.com/18611095/116218052-d0f7f500-a752-11eb-8d68-5aafc472d11b.png)

I use scxml in industrial devices in 24/7, so I must be sure that nothing new appers. Official uscxml was in a race to win additional speed. It was ok at that time when I was forked. I was much more interested to have stable and free from memory leaks library than have additional nanoseconds of speed.

If you are using uscxml in continious working devices, be careful of [this issue](https://github.com/tklab-tud/uscxml/issues/179)

**Comment by alexzhornyak** [comment]
Final patch version. All tests are passed. But I'm not sure that the logic is 100% correct

[LargeMicroStep.cpp.txt](https://github.com/alexzhornyak/uscxml/files/6384040/LargeMicroStep.cpp.txt)
